[Feature] Show current PR on status bar

jcansdale opened the pull request:
The following PRs target this one:

- Expose IVSGitExt as async VS service: #1510 (extends closed PR #1506)
- Use custom UIContext to load GitHubPackage / PullRequestStatusBarPackage in repo context: #1512
- Simplify the VSGitExt service: #1513

### Functionality

- Show PR associated with current branch on status bar
- Click on PR to navigate to PR details

![image](https://user-images.githubusercontent.com/11719160/35064847-3a94eb60-fbc3-11e7-92b7-1a6c129d5562.png)

- Tooltip shows PR number and title

![image](https://user-images.githubusercontent.com/11719160/36034111-e3ca828c-0daa-11e8-8ad0-d93305a925e7.png)

### To do

- [x] Make UI consistent with other VS status bar buttons (paging @donokuda)
- [x] Fix issue with PR number/title not updating
- [x] Add metrics for navigate to PR command
- [x] Make sure this feature isn't slowing startup
  - [x] Async load package
  - [x] Load package on background thread
  - [x] Load on `[ProvideAutoLoad(Guids.GitSccProviderId)]`?
  - [ ] ~~Use PR number from branch metadata?~~
  - [x] Lazy load `IPullRequestSessionManager`, see  https://github.com/github/VisualStudio/pull/1388/files#diff-8f8b4cd719001e4a8dd4b96e211243a8R53

### Related issues

- Show PR branch on status bar and allow navigation to PR details: #1102
- Highlight checked out PR in list: #1106

jcansdale commented:
@donokuda I've updated it to surface the number of submodules to sync. It currently looks like this:

![image](https://user-images.githubusercontent.com/11719160/35160116-236ffc0a-fd34-11e7-8991-2aca71aa5665.png)

meaghanlewis commented:
I've been dogfooding this for a couple days now. It's working well and allows me to remember what PR I'm looking at and to easily navigate back to that PR window.

One thing I find interesting is that when I checkout master, the PR I was just looking at still shows on the status bar and I didn't expect to see it there any longer.

shana commented:
> One thing I find interesting is that when I checkout master, the PR I was just looking at still shows on the status bar and I didn't expect to see it there any longer.

@jcansdale This feels like a bug?

jcansdale commented:
@shana,

> This feels like a bug?

Yes this is a bug. It's being tracked by #1440.

jcansdale commented:
@meaghanlewis You should find this (leaving a PR branch) fixed now. I've merged in #1442 to this PR.

meaghanlewis commented:
@jcansdale this is looking good to me now 👍

jcansdale commented on src/GitHub.InlineReviews/PullRequestStatusPackage.cs:
```diff
@@ -0,0 +1,23 @@
+    [PackageRegistration(UseManagedResourcesOnly = true)]
+    [Guid(Guids.PullRequestStatusPackageId)]
+    [ProvideAutoLoad(Guids.GitSccProviderId)]
+    public class PullRequestStatusPackage : Package
```
This shold be an aync package.

grokys commented:
The tooltip as it is could be slightly confusing:

![image](https://user-images.githubusercontent.com/1775141/36029689-87b9c6ca-0da4-11e8-84d0-a8a1320ca012.png)

This could be taken to mean "clicking this will rename the project dir". If it showed "#43 - Rename project dir" it might be clearer?

jcansdale commented:
@grokys Yup, I think that works.

![image](https://user-images.githubusercontent.com/11719160/36033902-20c6e50a-0daa-11e8-86f4-aa7667b3ff27.png)

What do you reckon?

donokuda commented:
Just driving by to say that I'm 👍 for adding the pull request number. I'm excited for this to land too! 😄

shana commented on src/GitHub.InlineReviews/PullRequestStatusBarPackage.cs:
```diff
@@ -0,0 +1,29 @@
+        {
+            var componentModel = (IComponentModel)await GetServiceAsync(typeof(SComponentModel));
+            var exportProvider = componentModel.DefaultExportProvider;
+            var pullRequestStatusManager = exportProvider.GetExportedValue<IPullRequestStatusBarManager>();
```
MEF calls need to happen on the main thread, and there's no guarantee that this method is going to be called on the main thread, especially once the new auto-load changes come (https://docs.microsoft.com/en-us/visualstudio/extensibility/how-to-use-asyncpackage-to-load-vspackages-in-the-background). For things that are required during package loading, they should be registered as services instead to avoid having to hit MEF on startup (see GitHubPackage.cs)

jcansdale replied:
I've changed `IVSGitExt` to be a VS services with no MEF dependencies.

grokys commented on src/GitHub.InlineReviews/PullRequestStatusBarPackage.cs:
```diff
@@ -0,0 +1,29 @@
+    [PackageRegistration(UseManagedResourcesOnly = true)]
+    [Guid(Guids.PullRequestStatusPackageId)]
+    [ProvideAutoLoad(Guids.GitSccProviderId)]
+    public class PullRequestStatusBarPackage : AsyncPackage
```
Does this need to be in a separate package? For context, I've been thinking recently that we should be merging `GitHub.InlineReviews` into our main package if possible, as it's completely dependent on it. Could this go in the main package? Would there be problems with that?

jcansdale replied:
When I first wrote this functionality, I made it piggyback a ride with one of the other packages. Keeping it separate can be useful to control exactly when and how it loads.

For example, I wanted this package load when the `GitSccProvider` is available and with the option `AllowsBackgroundLoading = true`. The package it was being loaded with before was using different settings (being loaded when a solution was opened).

shana replied:
The `GitHubPackage` already loads at the correct time in the correct way, btw.

jcansdale replied:
BTW, the `GitHubPackage` package uses `AllowsBackgroundLoading = true`, but we'd forgotten to use the `PackageAutoLoadFlags.BackgroundLoad` option on `[ProvideAutoLoad(...)]`. This means it uses a background thread when an executing command triggers the package load but the Main thread when the package is auto-loaded. Subtle and confusing! 😕 I've addresses this issue in #1512.

grokys commented:
Small thing, but I've just noticed that the color for this PR icon is different to the colors for the rest of the git status bar icons:

![image](https://user-images.githubusercontent.com/1775141/36125520-908d3338-1054-11e8-8387-2cba4284f7de.png)

donokuda commented:
> Small thing, but I've just noticed that the color for this PR icon is different to the colors for the rest of the git status bar icons:

mmm, I think I'm gonna have to snoop around for the correct visual studio color for that icon. 🤔 🔍

jcansdale commented:
@shana,

> MEF calls need to happen on the main thread, and there's no guarantee that this method is going to be called on the main thread, especially once the new auto-load changes come (https://docs.microsoft.com/en-us/visualstudio/extensibility/how-to-use-asyncpackage-to-load-vspackages-in-the-background). For things that are required during package loading, they should be registered as services instead to avoid having to hit MEF on startup (see GitHubPackage.cs)

I've converted the package to load on a background thread. It now uses the following attributes.

```
    [PackageRegistration(UseManagedResourcesOnly = true, AllowsBackgroundLoading = true)]
    [ProvideAutoLoad(Guids.GitSccProviderId, PackageAutoLoadFlags.BackgroundLoad)]
```

I've fixed the `VSGitExt` and `PullRequestStatusBarManager` services so that they can be loaded on the background thread. The other services that is depends on will be loaded just in time on the Main thread.

I started converting them to be VS services, but `VSGitExt` was problematic because it's one of those `GitHub.TeamFoundation` MEF services with two implementations. I've updated `VSGitExt` so that it can be loaded and used from a background thread. Visual Studio's `SComponentModel` is also now an async service.

jcansdale commented:
Still needs more work before `AllowsBackgroundLoading = true` works 100%. 😢

shana commented on src/GitHub.InlineReviews/PullRequestStatusBarPackage.cs:
```diff
@@ -0,0 +1,35 @@
+            var usageTracker = (IUsageTracker)await GetServiceAsync(typeof(IUsageTracker));
+            var serviceProvider = (IGitHubServiceProvider)await GetServiceAsync(typeof(IGitHubServiceProvider));
+            var componentModel = (IComponentModel)await GetServiceAsync(typeof(SComponentModel));
+            var pullRequestSessionManager = componentModel.DefaultExportProvider.GetExport<IPullRequestSessionManager>();
```
MEF calls need to happen on the main thread.

Also, it would be better if everything goes through `GitHubServiceProvider`. This way we can mock all these types (and even bypass MEF entirely if we want to) without the caller having to change any calls. `GitHubServiceProvider` already does look ups via ComponentModel, and the MEF lookups ensure that the results are loaded from our DLLS and not any other type that might be overriding our stuff, so that's a plus.

Ideally, though, we don't use MEF in package initialization, because loading the catalog for the first time is really expensive. Since you need to grab `IVSGitExt` here, I suggest you expose it as a service in the `ServiceProviderPackage` class like we do the other services like metrics, that way you'll get an instance directly with no need for MEF scanning everything.

shana replied:
Of course, if `IPullRequestSessionManager` has too many dependencies to be exposed as a service, we can go through MEF, but should always do it via the `GitHubServiceProvider` anyway.

jcansdale replied:
I've fixed it so that everything goes through `GitHubServiceProvider`. No more explicit dependency on `IComponentModel`.

jcansdale commented on src/GitHub.InlineReviews/PullRequestStatusBarPackage.cs:
```diff
@@ -0,0 +1,31 @@
+            var usageTracker = (IUsageTracker)await GetServiceAsync(typeof(IUsageTracker));
+            var serviceProvider = (IGitHubServiceProvider)await GetServiceAsync(typeof(IGitHubServiceProvider));
+
+            await ThreadingHelper.SwitchToMainThreadAsync();
```
There seems to be an issue where this await doesn't return until a solution is loaded. This means that when Visual Studio is opened without a solution, the PR list might be visible but the current PR doesn't appear on the status bar.

This isn't necessarily a problem, but I'd like to understand what's going on. I wonder if this will impact any of our other async loading packages? We wouldn't notice the issue on packages that use `[ProvideAutoLoad(UIContextGuids80.SolutionExists)]` (because they only load when a solution exists).

shana replied:
Yeah, this is yet another reason why we shouldn't be switching to the main thread for initialization whenever we can avoid it. If all we need are non-MEF services, we shouldn't need to switch to the main thread for any of this.

shana commented:
> I agree it would be best to expose IVSGitExt as a service. I'm inclined to do this as a separate PR though. What do you reckon?

I'd rather not introduce MEF loading code in the package if we're just throwing it away later. Maybe do a separate PR targeting this branch with that change?

StanleyGoldman commented:
![image](https://user-images.githubusercontent.com/417571/36736081-af02dd70-1ba5-11e8-86ac-cc2e18b3d5d8.png)

Playing with this locally, looks good.

jcansdale commented on src/GitHub.InlineReviews/Services/PullRequestStatusBarManager.cs:
```diff
@@ -0,0 +1,169 @@
+            this.usageTracker = usageTracker;
+            this.serviceProvider = serviceProvider;
+
+            OnActiveRepositoriesChanged();
```
This might be called twice!

jcansdale replied:
This will be fixed in #1512.

shana commented on src/GitHub.TeamFoundation.14/Services/VSGitExt.cs:
```diff
@@ -8,91 +9,67 @@
-        public VSGitExt(IGitHubServiceProvider serviceProvider)
-            : this(serviceProvider, new VSUIContextFactory(), new LocalRepositoryModelFactory())
+        public VSGitExt(Func<Type, Task<object>> getServiceAsync)
+            : this(getServiceAsync, new VSUIContextFactory(), new LocalRepositoryModelFactory())
```
Any reason to send a callback instead of sending in `IAsyncServiceProvider` (which provides `GetServiceAsync` here?

jcansdale replied:
The issue was that `AsyncPackage` implements 3 interfaces called `IAsyncServiceProvider`, 2 of which exist in the same namespace. Passing a delegate way a way to avoid taking a dependency on `Microsoft.VisualStudio.Shell.Immutable.14.0` and needing to disambiguate using `extern alias` (which works but is a bit messy).

I've pushed a commit that uses the interface from `Microsoft.VisualStudio.Shell.Immutable.14.0`. If we're likely to be using `IAsyncServiceProvider` elsewhere, having the correct assembly to hand is maybe a good thing (even if we need to use `extern alias`).

An alternative approach would be to use the lower level Interop `IAsyncServiceProvider` interface from here:

```
namespace Microsoft.VisualStudio.Shell.Interop
{
    public interface IAsyncServiceProvider
    {
        IVsTask QueryServiceAsync([ComAliasName("OLE.REFGUID")] ref Guid guidService);
    }
}
```

The fact that is takes a `Guid` and returns a `IVsTask` again makes it a bit messy.

My instinct is to go with this last commit. What do you think?

shana commented:
> The issue was that AsyncPackage implements 3 interfaces called IAsyncServiceProvider, 2 of which exist in the same namespace. Passing a delegate was a way to avoid taking a new dependency on Microsoft.VisualStudio.Shell.Immutable.14.0 and needing to disambiguate using extern alias (which works but is a bit messy).

Which two assemblies provide the same type?

jcansdale commented:
@shana,

It's a collision between `Microsoft.VisualStudio.Shell.Immutable.14.0` and `Microsoft.VisualStudio.Shell.Framework, Version=15`.

```
1>------ Build started: Project: GitHub.TeamFoundation.15, Configuration: Debug Any CPU ------
1>C:\Source\github.com\github\VisualStudio\src\GitHub.TeamFoundation.14\Services\VSGitExt.cs(15,60,15,81): error CS0433: The type 'IAsyncServiceProvider' exists in both
'Microsoft.VisualStudio.Shell.Framework, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' and 
'Microsoft.VisualStudio.Shell.Immutable.14.0, Version=14.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
========== Build: 0 succeeded, 1 failed, 20 up-to-date, 0 skipped ==========
```

I'll see what happens if I remove the `Microsoft.VisualStudio.Shell.Framework, Version=15` dependency from `GitHub.TeamFoundation.15`...

jcansdale commented:
Ah, we're forced to reference `Microsoft.VisualStudio.Shell.Framework, Version=15` because of this line. 😭 

![image](https://user-images.githubusercontent.com/11719160/36981049-fa70f840-2083-11e8-8c64-b2912673d9ac.png)

This would however be an easy reflection call.

jcansdale commented:
I've changed it so that rather than giving an alias to the shared `Microsoft.VisualStudio.Shell.Immutable.14` assembly that contains the `IAsyncServiceProvider` implementation that we want, we give an alias to `Microsoft.VisualStudio.Shell.Framework, v15` that we only use once in `VSGitServices`.

This will make finding the correct `IAsyncServiceProvider` a lot easier in future!
